receiver/pulsarreceiver/pulsar_receiver.go

```diff
@@ -0,0 +1,343 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulsarreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/pulsarreceiver"
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/zap"
+)
+
+var errUnrecognizedEncoding = errors.New("unrecognized encoding")
+
+const alreadyClosedError = "AlreadyClosedError"
+
+type pulsarTracesConsumer struct {
+	id              config.ComponentID
+	tracesConsumer  consumer.Traces
+	topic           string
+	client          pulsar.Client
+	cancel          context.CancelFunc
+	consumer        pulsar.Consumer
+	ready           chan error
+	unmarshaler     TracesUnmarshaler
+	settings        component.ReceiverCreateSettings
+	consumerOptions pulsar.ConsumerOptions
+}
+
+func newTracesReceiver(config Config, set component.ReceiverCreateSettings, unmarshalers map[string]TracesUnmarshaler, nextConsumer consumer.Traces) (*pulsarTracesConsumer, error) {
+	unmarshaler := unmarshalers[config.Encoding]
+	if nil == unmarshaler {
+		return nil, errUnrecognizedEncoding
+	}
+
+	options := config.clientOptions()
+	client, err := pulsar.NewClient(options)
+	if err != nil {
+		return nil, err
+	}
+
+	consumerOptions, err := config.consumerOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	return &pulsarTracesConsumer{
+		id:              config.ID(),
+		tracesConsumer:  nextConsumer,
+		topic:           config.Topic,
+		unmarshaler:     unmarshaler,
+		settings:        set,
+		client:          client,
+		ready:           make(chan error),
+		consumerOptions: consumerOptions,
+	}, nil
+}
+
+func (c *pulsarTracesConsumer) Start(context.Context, component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancel = cancel
+
+	go func() {
+		err := consumerTracesLoop(ctx, c)
+		if err != nil {
+			c.settings.Logger.Error("Consume tracesConsumer failed", zap.Error(err))
+		}
+	}()
+
+	err := <-c.ready
+	close(c.ready)
+	return err
+}
+
+func consumerTracesLoop(ctx context.Context, c *pulsarTracesConsumer) error {
+	unmarshaler := c.unmarshaler
+	traceConsumer := c.tracesConsumer
+
+	_consumer, err := c.client.Subscribe(c.consumerOptions)
+	c.ready <- err
+	if nil != err {
+		return err
+	}
```

Comment:
Why `c.client.Subscribe` needs to be called asynchronously instead of just setting `c.consumer` inside `Start`?

Reply:
@dmitryax it makes sense, I change it. PTAL